File: Week02/d1_sparse_matrix_transpose.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct {
    int row, col, val;
} term;

term* generate_matrix(int, int);
void display_matrix(term*);
term* transpose_matrix(term*);

int main() {
    srand(time(NULL));
    int row = rand() % 10 + 1, col = rand() % 10 + 1;
    term *a = generate_matrix(row, col);
    term *b = transpose_matrix(a);
    
    display_matrix(a);
    display_matrix(b);
    
    free(a);
    free(b);
    
    return 0;
}

term* generate_matrix(int row, int col) {
    printf("Generating Matrix\n");
    int i, j, r, idx = 0, n = row * col;
    term *m = (term*)malloc(sizeof(term) * (n + 1));
    
    m[0].row = row;
    m[0].col = col;
    m[0].val = 0;

    for (i = 0; i < row; i++) {
        for (j = 0; j < col; j++) {
            r = rand() % n;
            if (r < (n / 4)) {
                r = rand() % 90 + 10;
                m[++idx].row = i;
                m[idx].col = j;
                m[idx].val = r;
                m[0].val++;
            }
        }
    }

    m = (term*)realloc(m, sizeof(term) * (idx + 1));
    
    return m;
}

void display_matrix(term *m) {
    int idx = 1;
    printf("=== %dx%d Total %d ===\n", m[0].row, m[0].col, m[0].val);
    
    for (int i = 0; i <= m[0].val; i++)
        printf("(%d, %d, %d)\n", m[i].row, m[i].col, m[i].val);
    printf("\n");
    
    for (int i = 0; i < m[0].row; i++) {
        for (int j = 0; j < m[0].col; j++) {
            int val = 0;
            if (m[idx].row == i && m[idx].col == j)
                val = m[idx++].val;
            printf("%2d ", val);
        } printf("\n");
    } printf("\n");
}

term* transpose_matrix(term *a) {
    int i, j, idx = 1;
    term *b = (term*)malloc(sizeof(term) * a[0].val);

    b[0].row = a[0].col;
    b[0].col = a[0].row;
    b[0].val = a[0].val;

    for (i = 0; i < a[0].col; i++) {
        for (j = 1; j <= a[0].val; j++) {
            if (a[j].col == i) {
                b[idx].row = a[j].col;
                b[idx].col = a[j].row;
                b[idx].val = a[j].val;
                idx++;
            }
        }
    }

    return b;
}

File: Week02/d1_sparse_matrix_transpose.py
import random

def main():
    a = SparseMatrix(30, 50, rand=True)
    print(a, end='\n\n')
    a.transpose()
    print(a)

class SparseMatrix:
    def __init__(self, row, col, rand=False):
        self.row = row
        self.col = col
        self.terms = list()
        if rand:
            self._random_generate()
    
    def _random_generate(self):
        size = self.row * self.col
        n = size // random.randint(3, 5)
        
        m = [(i // self.col, i % self.col, random.randint(10, 99)) for i in range(size)]
        
        for i in range(n):
            r = random.randint(0, size - 1)
            t = m[r]
            m[r] = m[i]
            m[i] = t

        self.terms = sorted(m[:n])
        self.size = len(self.terms)

    def __str__(self):
        r = ['=== %dx%d Total %d ===' % (self.row, self.col, len(self.terms))]
        for term in self.terms:
            r.append('(%d, %d, %d)' % term)
        r.append('')
        
        idx = 0
        for i in range(self.row):
            col_str = []
            for j in range(self.col):
                val = 0
                if idx < self.size:
                    if i == self.terms[idx][0] and j == self.terms[idx][1]:
                        val = self.terms[idx][2]
                        idx += 1
                col_str.append('%2d' % val)
            r.append(' '.join(col_str))
        
        return '\n'.join(r)

    def transpose(self):
        terms = list()
        for term in self.terms:
            terms.append((term[1], term[0], term[2]))
        self.row, self.col = self.col, self.row
        self.terms = sorted(terms)

if __name__ == "__main__":
    main()